src/backend/distributed/master/master_citus_tools.c

```diff
@@ -0,0 +1,538 @@
+}
+
+
+/*
+ * ExecuteCommandsInParellelAndStoreResults connects to each node specified in
+ * nodeNameArray and nodePortArray, and executes command in commandStringArray
+ * in parallel fashion. Execution success status and result is reported for
+ * each command in statusArray and resultStringArray. Each array contains
+ * commandCount items.
+ */
+static void
+ExecuteCommandsInParallelAndStoreResults(StringInfo *nodeNameArray, int *nodePortArray,
+										 StringInfo *commandStringArray,
+										 bool *statusArray, StringInfo *resultStringArray,
+										 int commmandCount)
+{
+	int commandIndex = 0;
+	char *nodeUser = CurrentUserName();
+	PGconn **connectionArray = palloc0(commmandCount * sizeof(PGconn *));
+	int completedCount = 0;
+
+	/* establish connections */
+	for (commandIndex = 0; commandIndex < commmandCount; commandIndex++)
+	{
+		char *nodeName = nodeNameArray[commandIndex]->data;
+		int nodePort = nodePortArray[commandIndex];
+		PGconn *connection = ConnectToNode(nodeName, nodePort, nodeUser);
+		StringInfo queryResultString = resultStringArray[commandIndex];
+
+		statusArray[commandIndex] = true;
+
+		connectionArray[commandIndex] = connection;
+
+		if (connection == NULL)
+		{
+			appendStringInfo(queryResultString, "failed to connect to %s:%d", nodeName,
+							 (int) nodePort);
+			statusArray[commandIndex] = false;
+			completedCount++;
+		}
+	}
+
+	/* send queries at once */
+	for (commandIndex = 0; commandIndex < commmandCount; commandIndex++)
+	{
+		int querySent = 0;
+		PGconn *connection = connectionArray[commandIndex];
+		char *queryString = commandStringArray[commandIndex]->data;
+		StringInfo queryResultString = resultStringArray[commandIndex];
+
+		/*
+		 * If we don't have a connection, nothing to send, error string should already
+		 * been filled.
+		 */
+		if (connection == NULL)
+		{
+			continue;
+		}
+
+		querySent = PQsendQuery(connection, queryString);
+
+		if (querySent == 0)
+		{
+			StoreErrorMessage(connection, queryResultString);
+			statusArray[commandIndex] = false;
+			PQfinish(connection);
+			connectionArray[commandIndex] = NULL;
+			completedCount++;
+		}
+	}
+
+	/* check for query results */
+	while (completedCount < commmandCount)
+	{
+		for (commandIndex = 0; commandIndex < commmandCount; commandIndex++)
+		{
+			PGconn *connection = connectionArray[commandIndex];
+			StringInfo queryResultString = resultStringArray[commandIndex];
+			bool success = false;
+			bool queryCompleted = false;
+
+			if (connection == NULL)
+			{
+				continue;
+			}
+
+			queryCompleted = GetConnectionStatusAndResult(connection, &success,
+														  queryResultString);
+
+			if (queryCompleted)
+			{
+				completedCount++;
+				statusArray[commandIndex] = success;
+				connectionArray[commandIndex] = NULL;
+				PQfinish(connection);
+			}
+		}
+
+		if (completedCount < commmandCount)
+		{
+			long sleepIntervalPerCycle = RemoteTaskCheckInterval * 1000L;
+			pg_usleep(sleepIntervalPerCycle);
+		}
```

Comment:
OK, however PQExec() still blocks. Hitting ctrl-c does not do anything while it is running.